lib/mayaUsd/ufe/UsdAttributes.cpp

```diff
@@ -127,26 +124,22 @@ Ufe::Attribute::Type UsdAttributes::attributeType(const std::string& name)
     auto iter = fUsdAttributes.find(name);
     if (iter != std::end(fUsdAttributes))
         return iter->second->type();
-    PXR_NS::TfToken tok(name);
+
+        // Shader definitions always win over created UsdAttributes:
 #ifdef UFE_V4_FEATURES_AVAILABLE
 #if (UFE_PREVIEW_VERSION_NUM >= 4010)
-    Ufe::NodeDef::Ptr nodeDef = UsdAttributes::nodeDef();
-    if (nodeDef) {
-        Ufe::AttributeDef::ConstPtr     attrDef = nameToAttrDef(tok, nodeDef);
-        UsdShaderAttributeDef::ConstPtr shaderAttrDef
-            = std::dynamic_pointer_cast<const UsdShaderAttributeDef>(attrDef);
-        if (shaderAttrDef) {
-            const auto& shaderProperty = shaderAttrDef->shaderProperty();
-            if (shaderProperty) {
-                return usdTypeToUfe(shaderProperty);
-            }
-        }
+    PXR_NS::SdrShaderPropertyConstPtr shaderProp = _GetSdrPropertyAndType(fItem, name).first;
+    if (shaderProp) {
+        return UsdShaderAttributeDef(shaderProp).type();
```

Comment:
We can still transiently create a Ufe/USD class when we want to reuse a good API. These classes are created on the stack and can be heavily optimized in release.